internal/errors/errors.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package errors
+
+import (
+	stderrors "errors"
+)
+
+// annotated is a wrapping error type that allows an already established error
+// to be annotated with another error so that the new annotated error now
+// satisfies both the original error and the annotation error with respect to
+// Is() and As()
+//
+// annotated only implements stderrors.Unwrap() []error returning both errors
+// involved in the annotation. This means that annotated errors are not
+// injecting the new annotation into the error chain and calls to errors.Unwrap
+// will never return the annotation error.
+type annotated struct {
+	error
+	annotation error
+}
+
+// Error provides a way to enrich an already existing Go error and inject new
+// information into the errors chain. Error and its operations are all immutable
+// to the encapsulated error value.
+type Error interface {
+	// error is the error being wrapped.
+	error
+
+	// Add will introduce a new error into the error chain so that subsequent
+	// calls to As() and Is() will be satisfied for this added error. The error
+	// being added here will not appear in the error output from Error().
+	// Unwrap() does not unwrap errors that have been added.
+	Add(err error) Error
+
+	// Trace records the location of where this function was called on the
+	// Error. Making it so that subsequent calls to Trace() produce a stack of
+	// traced errors.
+	//
+	// ErrorStack() can be used to produce a string based stack of all the
+	// errors in a change and any subsequent tracing information associated with
+	// the error.
+	Trace() Error
+
+	// Unwrap returns the underlying error being enriched by this interface.
+	Unwrap() error
+}
+
+// link is an implementation of Error and represents another link in a chain of
+// errors. It provides a way to wrap an existing error and offer functions to
+// further enrich the error chain with new information.
+//
+// link wants to be completely transparent to the error chain. All std errors
+// introspection on link works on the underlying error being wrapped.
+type link struct {
+	// error is the underlying error being wrapped by link. We use error in
+	// composition here so that the errors implementation of the interface makes
+	// link conform as an error type.
+	error
+}
+
+// Add implements Errors.Add.
+func (l link) Add(err error) Error {
+	if err == nil {
+		return l
+	}
+	return link{annotated{l.error, err}}
+}
+
+//func (l link) As(target any) bool {
+//	if l.error == nil {
+//		return false
+//	}
+//	if reflect.TypeOf(l.error).AssignableTo(reflect.TypeOf(target)) {
+//		reflect.ValueOf(target).Elem().Set(reflect.ValueOf(l.error))
+//		return true
+//	}
+//	return false
+//}
+
+func (l link) Is(target error) bool {
+	return l.error == target
+}
+
+// Trace implements Errors.Trace.
+func (l link) Trace() Error {
+	return link{newTraced(l.error, 1)}
+}
```

Comment:
Do we want this opt-in or opt-out? Currently, our package is opt-in, if we want to make this change we should socialize this change. When do we want to trace the error, at a boundary layer or every error?

Reply:
Let's discuss this today. Personally I have never like it but I am happy to keep it as the changes are worth it.

Reply:
The outcome from the conversations today.

- We are going to trace all newly created errors.
- Traced errors will store the program counter instead of the location.
- Trace() can be removed from Error as errors are traced by default.